### DISCUSSION TITLE
Add support for "saved searches" (i.e. former Quick Links GMail enhancement)

### DISCUSSION THREAD

**Discussion by cipriancraciun** [zoidsh/meru#250] [Feature Requests] [answered]
Long time-ago (before Google decide to decimate the GMail Labs ecosystem), there was an official "Quick Links" extension that allowed one to save searches.  Unfortunately I'm not able to find the original documentation but the following pages can provide some context:

* https://www.reddit.com/r/google/comments/6rkv8k/quick_links_lab_feature_in_gmail_is_retiring/
* https://www.bettercloud.com/monitor/the-academy/quick-links-lab-for-gmail-2/

So, this feature was priceless for many advanced GMail-ers (myself included), as it provided something that you couldn't obtain with GMail labels and other options.  Basically it offered a kind of "views" for your GMail account.  For example my inbox was always empty, all emails being properly tagged, and I leveraged the saved searches to create multiple "inboxes".

Thus my proposal is to add such an option, even in the most minimal way possible.  (I don't even mind editing a configuration file to get access to them.)  Basically it would just require bookmarking the search URL's.

**Comment by mskelton** [comment]
Could this be accomplished using Gmail filters?

**Comment by cipriancraciun** [comment]
> Could this be accomplished using Gmail filters?

What I'm referring to is orthogonal to GMail filters.

Say for example that your workflow includes reviewing unread emails received in muted threads.  You would achieve this by searching for `is:muted is:unread`.

However the same could be achieved by opening the following link: https://mail.google.com/mail/u/0/#search/is%3Amuted+is%3Aunread

Basically the Quick Links extension allowed one to bookmark any link that belonged to `https://mail.google.com/mail/...` and provided the user quick access to it.  That's why I'm saying that implementing this could be quite trivial, by only needing to replace the URL, as GMail, even in the browser has support for this.  (Just paste that URL in the same tab you have GMail open, and you'll see that you'll get the same outcome as manually typing the search query.)

**Comment by timche** [comment]
Loading URLs would require a full reload which isn't great for the experience. We could instead automatically insert the filters (`is:muted`, etc.) into the search bar and submit. I don't know how many possible filters exist, but I think the "search filter" could be configured by the user using the editor (file format tbd) and we could either inject a menu into the page, set shortcuts and/or add a new menu item.

**Comment by cipriancraciun** [comment]
> Loading URLs would require a full reload which isn't great for the experience.

No it wouldn't.  Just try it in a normal browser:
* open your GMail account;
* select the URL bar and paste the link containing the search query;

**Comment by timche** [comment]
Alright, it works, at least in Chrome. We need to check if it behaves the same in Electron. From an implementation perspective it wouldn't change much except loading an URL instead of injecting into the search bar.

**Comment by cipriancraciun** [comment]
If you implement this as a "bookmark" facility, one could also open other parts of the GMail application in a similar fashion.  Like for example: https://mail.google.com/mail/u/0/#settings/filters -- for quick access to filters;

**Comment by cipriancraciun** [comment]
Any updates on this item?  (As with #95, this is the only reason I still have to use sometimes GMail from a browser.)

**Comment by timche** [comment]
We can implement this, but I'm wondering how this feature should look like and how it's accessible.

**Comment by cipriancraciun** [comment]
The UI part is indeed the tricky part...  There are at least two distinct problems:
* how to manage these bookmarked searches;  (i.e. add, delete, re-order, etc.)
* how to display these bookmarks in the GMail UI.

Given that this would be an advanced feature I would say that management could be done by simply instructing the user to edit the `config.json` file and allow him to enter a list of bookmarks as for example:
~~~~
"bookmarks" : [
    {"label" : "muted", "search" : "is:muted is:unread"},
    {"label" : "filters", "href" : "#settings/filters"}
]
~~~~

As for the syntax:
* if the object contains a `search` key it means that the value is the query to be put in the search bar;
* if the object contains a `href` key, then as described in previous items it can just be loaded by replacing the URL of the window to give access to filters and other quick shortcuts;  (the value should begin with `#`, to make sure one isn't escaping GMail's UI;)


As for the actual UI, I would imagine a button injected somewhere in the GMail UI, that would open a list of bookmarks.  (Perhaps a simpler approach would be an option in the menu bar, but if one uses hidden menu bars, the visual effect would be an issue.)

**Comment by timche** [comment, marked as answer]
With v3 (#242) we have better possibilities to integrate this feature and I can see this being adding soon. So stay tuned!
